mat_dp_pipeline/pipeline/flatten_hierarchy.py

```diff
@@ -0,0 +1,131 @@
+import logging
+from collections import defaultdict
+from pathlib import Path
+from typing import Iterator
+
+import pandas as pd
+
+from mat_dp_pipeline.common import Tree, create_path_tree, validate
+from mat_dp_pipeline.pipeline.common import SparseYearsInput
+from mat_dp_pipeline.sdf import StandardDataFormat, Year
+
+
+def stringify_tree(tree: Tree, collapse_limit: int | None = None) -> list[str]:
+    def add_node_to_result(node: Tree, result: list[str], prefix: str):
+        """
+        Adds the specified node to the result list, via a depth first search.
+        """
+        if node is None:
+            return None
+
+        for i, (k, v) in enumerate(node.items()):
+            if i == 0:
+                new_prefix = f"{prefix} -> {k}"
+            else:
+                new_prefix = f"{len(prefix)*' '} -> {k}"
+            if v is None:
+                result.append(new_prefix)
+            elif len(v) > (collapse_limit or float("inf")):
+                result.append(new_prefix + f" -> ({len(v)} items)")
+            else:
+                add_node_to_result(v, result, new_prefix)
+
+    result = []
+    add_node_to_result(tree, result, "")
+
+    return result
+
+
+def overlay_in_order(
+    df: pd.DataFrame,
+    base_overlay: pd.DataFrame,
+    yearly_overlays: dict[Year, pd.DataFrame],
+) -> pd.DataFrame:
+    overlaid = df
+
+    # Overlays sorted by year, first one being base_overlay (year 0)
+    sorted_overlays = sorted(({Year(0): base_overlay} | yearly_overlays).items())
+
+    for year, overlay in sorted_overlays:
+        if overlay.empty:
+            continue
+
+        # Add "Year" level to the index. Concat is idiomatic way of doing it
+        update_df = pd.concat({year: overlay}, names=["Year"])
+
+        if overlaid.empty:
+            overlaid = update_df.copy()
+        else:
+            overlaid = update_df.combine_first(overlaid)
+
+    return overlaid
+
+
+def flatten_hierarchy(
+    root_sdf: StandardDataFormat,
+) -> list[tuple[Path, SparseYearsInput]]:
+    def dfs(
+        root: StandardDataFormat, inpt: SparseYearsInput, label: Path
+    ) -> Iterator[tuple[Path, SparseYearsInput, set[str]]]:
```

Comment:
The sofas are back! :D Maybe
root -> sdf
inpt -> sparse_years
dfs -> add_sparse_years_to_sdf

Reply:
add_sparse_years_to_sdf? Doesn't sound very good... Doesn't even sound correct.